segmentstore/server/src/main/java/io/pravega/segmentstore/server/CacheManager.java

```diff
@@ -362,6 +384,34 @@ private void logCurrentStatus(CacheStatus status) {
                 status.getSize() / 1048576);
     }
 
+    private void notifyCleanupListeners() {
+        ArrayList<CacheUtilizationProvider.CleanupListener> toNotify = new ArrayList<>();
+        ArrayList<CacheUtilizationProvider.CleanupListener> toRemove = new ArrayList<>();
+        synchronized (this.cleanupListeners) {
+            for (CacheUtilizationProvider.CleanupListener l : this.cleanupListeners) {
+                if (l.isClosed()) {
+                    toRemove.add(l);
+                } else {
+                    toNotify.add(l);
+                }
+            }
+
+            this.cleanupListeners.removeAll(toRemove);
+        }
+
+        for (CacheUtilizationProvider.CleanupListener l : toNotify) {
+            try {
+                l.cacheCleanupComplete();
+            } catch (Throwable ex) {
+                if (Exceptions.mustRethrow(ex)) {
+                    throw ex;
+                }
+
+                log.error("{}: Error while notifying cleanup listener {}.", TRACE_OBJECT_ID, l, ex);
```

Comment:
No it's not unrecoverable and no action is required. I put try-catch here so that if one listener decides to throw an unexpected exception, we don't prevent other listeners from being notified. If there is an error, we should log it so that someone can take a look at why it got thrown in the first place.